sdks/python/apache_beam/utils/windowed_value.py

```diff
@@ -279,6 +293,208 @@ def create(value, timestamp_micros, windows, pane_info=PANE_INFO_UNKNOWN):
   return wv
 
 
+class BatchingMode(Enum):
+  CONCRETE = 1
+  HOMOGENEOUS = 2
+
+
+class WindowedBatch(object):
+  """A batch of N windowed values, each having a value, a timestamp and set of
+  windows."""
+  def with_values(self, new_values):
+    # type: (Any) -> WindowedBatch
+
+    """Creates a new WindowedBatch with the same timestamps and windows as this.
+
+    This is the fasted way to create a new WindowedValue.
+    """
+    raise NotImplementedError
+
+  def as_windowed_values(self, explode_fn: Callable) -> Iterable[WindowedValue]:
+    raise NotImplementedError
+
+  @staticmethod
+  def from_windowed_values(
+      windowed_values: Sequence[WindowedValue],
+      *,
+      produce_fn: Callable,
+      mode: BatchingMode = BatchingMode.CONCRETE) -> Iterable['WindowedBatch']:
+    if mode == BatchingMode.HOMOGENEOUS:
```

Comment:
Do we need an enum here, or should we just implement this on the subclasses, e.g. HomogeneousWindowedBatch.from_windowed_values to get a set of homogeneous windowed values?

Reply:
Maybe we could want to hide the concrete subclasses, but it seems like this always gets overridden. Perhaps we don't need the concrete implementation until we solve the questions of batching + windowing in the API.

Reply:
Yeah.. I think it make sense to just keep it simple and tear out Concrete(Heterogeneous)WindowedBatch for now.

Reply:
+1

Reply:
Done, [0f2e5b9](https://github.com/apache/beam/pull/17384/commits/0f2e5b9243e70eb6cf4f7c60f11c636371cad849)